### PR TITLE
Relocatable tar.gz for Ubuntu

### PR DESCRIPTION
**Pull request by heylel-b-sh**
**General info**
To build relocatable `tar.gz` (with a **PostreSQL** and **LocalExecutor**) that can be run with PM2 on Ubuntu 18.04 start clean virtual machine with Ubuntu 18.04
and run the following commands. Optionally, the default programs versions used in a script can be redefined if `.env` file is provided in the same folder as `build_ubuntu.sh` script.

```bash
sudo apt-get install git g++ make curl
curl https://install.meteor.com/ | sh                          # installing meteor with its own node for building BioWardrobe-NG
git clone https://github.com/datirium/scidap-satellite.git
cd ./scidap-satellite/
git checkout ubuntu_packaging                                  # this can be skipped when PR is merged
git pull origin ubuntu_packaging                               # this can be skipped when PR is merged
cd ./build-scripts
./build_ubuntu.sh
```

After script finished running, you will find a compressed `scidap-satellite.tar.gz` in the `../ubuntu_post_build` folder. All the temporary data is kept in `../build_ubuntu` and can be removed unless you want to save some time when rerunning `build_ubuntu.sh` script.

To run relocatable `scidap-satellite.tar.gz` on clean Ubuntu 18.04 with installed and configured Docker use the following commands.
```bash
sudo apt-get install git nodejs npm curl              # mongod doesn't work without curl
mkdir satellite
mv scidap-satellite.tar.gz satellite                  # initial location of scidap-satellite.tar.gz might be different
cd satellite
tar xzf scidap-satellite.tar.gz
npm install pm2
./node_modules/pm2/bin/pm2 start ./configs/ecosystem.config.js  # start ecosystem.config.js with PM2
```

Default configuration is saved in the `./configs/scidap_default_settings.json` file. Custom configuration can be loaded from the file set in `SCIDAP_SETTINGS` environment variable, `../scidap_settings.json` or `~/.config/scidap-satellite/scidap_settings.json`.

```yaml
{
    "defaultLocations": {
        "files": "files",
        "mongodb": "mongodb",
        "airflow": "airflow",
        "satellite": "satellite",
        "pgdata": "pgdata"
    },
    "satelliteSettings": {
        "rcServerToken": "",
        "rcServer": "https://api-sync.scidap.com",
        "port": 3069,
        "scidapRoot": "./scidap",
        "airflowAPIPort": 8080,
        "aria2cPort": 6800,
        "mongoPort": 27017,
        "mongoCollection": "scidap-satellite",
        "baseUrl": "http://localhost:3069/",
        "scidapSSLPort": 3070,
        "sslCert": "",
        "sslKey": "",
        "localFiles": true,
        "proxy": "",
        "noProxy": ""
    },
    "airflowSettings": {
        "core.executor": "LocalExecutor",
        "core.dag_concurrency": 2,
        "core.dags_are_paused_at_creation": "False",
        "core.load_examples": "False",
        "core.max_active_runs_per_dag": 1
    },
    "aria2cSettings": {
        "--console-log-level": "debug",
        "--remove-control-file": true,
        "--allow-overwrite": true,
        "--auto-file-renaming": false
    },
    "databaseSettings": {
        "db_port": 5432,
        "db_user": "airflow",
        "db_password": "airflow",
        "db_name": "airflow"
    },
    "meteorSettings": {
        "logLevel": "debug",
        "cors_package": true,
        "email": {
            "url": "",
            "from": ""
        },
        "extra_users": [],
        "public": {
        },
        "accounts": {
            "sendVerificationEmail": true,
            "forbidClientAccountCreation": true,
            "loginExpirationInDays": 7
        },
        "ldap": {},
        "oauth2server": {},
        "remotes": {
            "postform": {
                "collection": {},
                "publication": "none"
            },
            "directurl": {
                "caption": "Direct URL",
                "type": "urls",
                "protocol": [
                    "https",
                    "http",
                    "ftp"
                ],
                "refreshSessionInterval": 180
            },
            "geo": {
                "caption": "GEO",
                "type": "urls",
                "protocol": [
                    "geo"
                ],
                "refreshSessionInterval": 180
            }
        }
    }
}
```

**Additional details for developers**
1. Had to patch `cwltool/sandboxjs.py` - `cwltool/sandboxjs.py` fails to evaluate JavaScript expressions when CWL-Airflow is run from PM2. Perhaps, it's somehow related to how PM2 starts the project. Out of curiosity, tested if exactly the same command works with `child_process.spawnSync` and it worked. As a workaround, made `cwltool` always run JavaScript expression evaluation using Docker. In the future we won't have this problem, as all CWL-Airflow related parts will be moved in docker-compose.
2. Couldn't compile aria with `--with-openssl` even if openssl was installed. Decided not to spend much time and to use builds from here https://github.com/q3aql/aria2-static-builds
3. PostgeSQL binaries are downloaded from https://www.enterprisedb.com/download-postgresql-binaries 
4. We don’t need to think how to retrieve the token based on the username and password as this part is tightly bound to Meteor logic which is going to be replaced soon. So the current approach is that the user will go to his profile and manually copy his token for satellite.
5. In future put Airflow Scheduler, CWL-Airflow API and MySQL (optionally some other DB) into docker-compose start them all together from inside PM2. BioWardrobe-NG, MongoDB, Aria2c stay outside docker-compose. We need PM2 for remote access to logs and control through UI.
6. In future start PM2 as a service

**Remote monitoring**
Free account on https://id.keymetrics.io/api/oauth/register allows to track the limited number of processes (max 4). On the attached picture below we see that `airflow-scheduler` process is ignored because of this limitation. However, it looks like you can at least restart applications. Super easy to link you pm2 project to UI with `pm link` commands.

Limitation:
- Fetching logs took too long, so it's not properly tested.
- All cool dashboards and more complicated process controls require payed subscription.
- It seams like sometimes when all processes started from the UI their order is arbitrary and BioWarbdobe-NG is started before MongoDB
- Not checked how it works behind the firewall.
Screen Shot 2020-11-24 at 2 18 59 PM

**Possible risks when merging**
When trying to update the current scidap-satellite installation the possible problems might me related to not being able to catch up old mongo database with the newer version of MongoDB

**Upgrading from the old Ubuntu satellite installation**
1. Stop services
```bash
systemctl stop airflow-apiserver
systemctl stop airflow-scheduler
systemctl stop airflow-webserver
systemctl stop aria2c
systemctl stop scidap-satellite
```
2. Create dump from the current MongoDB and stop `mongodb` service
```
mongodump
systemctl stop mongodb
```
3. Start only `mongod` using PM2 `--only` parameter and restore database from the the dump
```
mongorestore
```
4. Start all services using PM2

**Known issues**
1. Can't find Node's modules
```export NODE_PATH="'$(npm root)'"```
or
```export NODE_PATH="'$(npm root -g)'"```
2. Can't start `airflow-scheduler` or `airflow-apiserver` services
    Make sure you don't have `~/.local` with python packages installed, especially with `cwl-airflow`. Quick solution - just rename it. Even if we set PYTHONPATH, it will still try to search in the standard python module locations.
3. Docker seems to works slow even on `docker run`. Check if you have the following line in the `journalctl -u docker` log
```
INFO[2021-01-03T12:21:16.363784091-05:00] No non-localhost DNS nameservers are left in resolv.conf. Using default external servers: [nameserver 8.8.8.8 nameserver 8.8.4.4]
```
Check if you can ping either 8.8.8.8 or 8.8.4.4. If not, probably you are using a private (corporate) DNS server. Add dns configuration into your `/etc/docker/daemon.json` file. Check what DNS settings you have in `/etc/resolv.conf` file
```
"dns": ["127.0.0.53", "8.8.8.8", "8.8.4.4"]
```
More details can be found https://docs.docker.com/engine/install/linux-postinstall/#specify-dns-servers-for-docker
and https://stackoverflow.com/a/39400887/8808721
Alternatively, test `docker run` with `--dns` param to see if DNS was the problem

**Building App for macOS**
On macOS with preinstalled `meteor`, `node` and `git` run the following commands
```bash
git clone https://github.com/datirium/scidap-satellite.git
cd ./scidap-satellite/
git checkout ubuntu_packaging                                 # this can be skipped when PR is merged
git pull origin ubuntu_packaging                               # this can be skipped when PR is merged
npm install
cd ./build-scripts
./build_macos.sh
cd ..
npm run electron:mac
```
In case you by mistake run `build_ubuntu.sh` script, you need to remove `node_modules`, `build`, `build_ubuntu`, `Services`, `ubuntu_post_build` folders in the root of you project and rerun `npm install`, `build_macos.sh` and `npm run electron:mac` commands as it's mentioned above.

The `scidap-satellite.app` will be saved in `./release/mac` folder.

After bundle update both for Ubuntu and macOS you might need to fetch new CWL. For that remove `CWL` field from the `sync` object of the correspondent document in the `Satellites` collection. **Do not** remove other fields as that will cause samples to rerun.

### CODE REVIEW

**Comment by heylel-b-sh** [comment]
**DONE**

Do not forget to update `port` from the `meteor_default_settings.json` if you set a different value for `aria2cPort` in `ecosystem.config.js`. Or make it updated automatically when generating meteor settings.
```
    "download": {
        "aria2": {
            "host": "localhost",
            "port": 6800,
            "secure": false,
            "secret": "",
            "path": "/jsonrpc"
        }
    }
```

**Comment by heylel-b-sh** [comment]
Note, we use `fs.mkdirSync(folder, {recursive: true})` but only NodeJS version 10.12.0 has added a native support for that flag.

**Comment by heylel-b-sh** [comment]
**DONE**

`triggerDag` should be automatically updated based on `airflowAPIPort` value

**Comment by heylel-b-sh** [comment]
Think if I need to add `--no-proxy=` for Aria2c

**Comment by heylel-b-sh** [comment]
Looks like we don't need `const os = require('os');` for ubuntu script

**Comment by heylel-b-sh** [comment]
**NOT RELEVANT**

Maybe for `airflowSettings` we don't need "False" as string. Check if it's ok for Airflow to have boolean in a json format (just simple false)

File: README.md [hunk 1]
```diff
@@ -39,3 +39,192 @@ Please follow [Angular-cli documentation](https://github.com/angular/angular-cli
...
+systemctl stop airflow-webserver
+systemctl stop aria2c
+systemctl stop scidap-satellite
+```
+2. Create dump from the current MongoDB and stop `mongodb` service
+```
+mongodump
+systemctl stop mongodb
+```
+3. Start only `mongod` using PM2 `--only` parameter and restore database from the the dump
```
Comment (portah) [inline README.md:204]: Would not be the total command be easier? `pm2 ... --only ...` ?

**Review by portah — DISMISSED** [review]
Have not caught any red flags...

**Review by portah — DISMISSED** [review]

File: README.md (hunk 1 above)
Comment (heylel-b-sh) [reply README.md:204]: Updated

**Review by dtroyano86 — APPROVED** [review]

**Pull request merged**